### PR TITLE
Absolute require

### LINKED ISSUE #43: Intercept absolute paths
It seems like it is impossible to use require-in-the-middle to intercept requires of absolute file paths. For instance, I would like to intercept the require of `/var/runtime/util.js` which is outside of my `node_modules` and not a core module. I tried to do `hook(['/var/runtime/util'], onRequire)`, but when `require('/var/runtime/util')` is called, require-in-the-middle logs `require-in-the-middle could not parse filename: /var/runtime/util.js +0ms`.

It looks like require-in-the-middle is using the module-details-from-path module, which fails to parse this absolute path.

### PR DESCRIPTION
**Pull request by dyladan**
Fixes elastic/require-in-the-middle#43

### CODE REVIEW

**Comment by dyladan** [comment]
I signed the CLA. How do I make the check pass?

File: index.js
Comment (watson) [inline index.js:103]: What edge-case does the 2nd `modules.includes()` cover? I notice that it doesn't include the file-extension (I'm not sure if this is on purpose).

File: index.js
Comment (watson) [inline index.js:63]: I prefer if we explicitly set `absoluteRequire` to `false` as the default value. That signals intent better and helps V8 😃

File: test/test.js
Comment (watson) [inline test/test.js:256]: It would be nice with an additional test to verify that `require('/foo/bar')` correctly triggers `hook(['/foo/bar/index.js'], ...)`

**Review by watson — COMMENTED** [review]
Looking good 👍 I just left a few comments

File: index.js
Comment (dyladan) [reply index.js:103]: If filename is `'/path/to/file.js'`, this turns into `modules.includes('/path/to/file.js') || modules.includes('/path/to/file')`.

It allows the user to specify or not the `.js` file extension

File: test/test.js
Comment (dyladan) [reply test/test.js:256]: I had not considered that use case. Added in 96e6b6bdd6c549b52e5634cf1ef0426e33852f68

File: index.js
Comment (dyladan) [reply index.js]: Added comments to clarify

Comment (watson) [reply index.js]: Ah ok, it's to allow the user to get the same experience no matter if they call `hook(['/a'], ...)` or `hook(['/a.js'], ...)`.

I had a couple of thoughts while thinking about this, all based on the following:

> `filename` is generated by `Module._resolveFilename(id, this)`, which should normalize all paths, so that both `/foo/bar/index` and `/foo/bar` becomes `/foo/bar/index.js` and that relative paths like `./foo` becomes `/path/to/foo.js` etc.

Thoughts:

- The `path.isAbolute(filename)` check above will always return `true` and can therefore be omitted.
- Since `filename` is always an absolute path, I think it's best if we made support for absolute paths something the user had to actively turn on using a config option. Otherwise one could argue if this was a breaking change. So the `path.isAbolute(filename)` check could be replaced with a check for this config option instead.
- Nit: I think it's a reasonable requirement to ask the user to include the file extension in absolute paths when calling `hook()`. That way we could do away with the 2nd call to `modules.include()`. Would this be annoying for the user you think?

Comment (watson) [reply index.js]: @dyladan any thoughts on this?

**Comment by dyladan** [comment]
Any update here on if this can be merged or if we need any more changes?

**Comment by watson** [comment]
@dyladan did you see my comment here: https://github.com/elastic/require-in-the-middle/pull/45#discussion_r476508561

File: index.js
Comment (dyladan) [reply index.js]: Oops I totally missed this.

>The path.isAbolute(filename) check above will always return true and can therefore be omitted.

Will remove

> Since filename is always an absolute path, I think it's best if we made support for absolute paths something the user had to actively turn on using a config option. Otherwise one could argue if this was a breaking change. So the path.isAbolute(filename) check could be replaced with a check for this config option instead.

If filename is always an absolute path, `modules.includes(filename)` will only be true if the user included an absolute path in the modules array. I think including an absolute path in the array sufficiently conveys the intent of the user that they would like to hook absolute paths, and I think a config option would be redundant. Unless I'm missing some use case where that assumption is incorrect?

>Nit: I think it's a reasonable requirement to ask the user to include the file extension in absolute paths when calling hook(). That way we could do away with the 2nd call to modules.include(). Would this be annoying for the user you think

I'm fine with making it so that absolute paths have to be truly absolute, including the file extension. This makes the change a little safer as well.

**Comment by dyladan** [comment]
> @dyladan did you see my comment here: [#45 (comment)](https://github.com/elastic/require-in-the-middle/pull/45#discussion_r476508561)

I'm sorry I missed it. Implemented those changes in 33b9ce0

**Comment by dyladan** [comment]
I have another quick question. If the user does `hook(['/absolute/path/node_modules/mod1/lib/file.js'] ...` in the currently implementation, the hook would receive the same basedir and name as if they hooked `hook(['mod1/lib/file'] ...` instead of the "absolute" style where the file name is the module name even though it is an absolute require. Is this behavior correct, or should we move the absolute require check above the `stat = parse(filename)` to prevent that?

**Comment by dyladan** [comment]
> I have another quick question. If the user does `hook(['/absolute/path/node_modules/mod1/lib/file.js'] ...` in the currently implementation, the hook would receive the same basedir and name as if they hooked `hook(['mod1/lib/file'] ...` instead of the "absolute" style where the file name is the module name even though it is an absolute require. Is this behavior correct, or should we move the absolute require check above the `stat = parse(filename)` to prevent that?

I pushed another change 6de600fd87012074c950a9b0bffa27ad19142847 which really makes this change very small. It brings the new absolute check above the `parse`. Because the absolute check is a simple `includes` on a short array, it should have a minimal performance impact.

**Comment by dyladan** [comment]
@watson can you please review the current state? It is quite a bit simpler than it was when you first reviewed.

**Comment by dyladan** [comment]
friendly bump

**Comment by dyladan** [comment]
@watson are you waiting on me for something? I have the feeling that we're waiting on each other here. :)

**Review by astorm — APPROVED** [review]
Sorry for the delay in getting back to you @dyladan, we've had some teams and responsibilities shuffle around a bit.  Also, nice to see you again, I think we interacted a bit on the oTel repo a while back :)

Overall this looks like a reasonable PR to me.  Checking if the resolved filename is the same as the filename that you're requesting to patch is a smart way to detect the client programmer's intent here, and it _shouldn't_ conflict with either the `module-name`, `module-name/foo.js` use cases, and the absolute paths will will still be detected as "not a core" module by `isCore`.  

Approving -- I'll ping @watson internally for a gut check before merging.  Thanks for your patience here. :)

**Comment by astorm** [comment]
Thanks for your patience and for adding this @dyladan -- PR's merged.

**Comment by Flarna** [comment]
Is is planned to publish a new version including this change?

**Comment by trentm** [comment]
@watson Any objections to my tagging/publishing a v5.1.0 release?

**Comment by watson** [comment]
@trentm published 👍

**Pull request merged**
